[CA-1224] Adds google project display name and tests

s-rubenstein opened the pull request:
related to: https://github.com/broadinstitute/workbench-libs/pull/597

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

s-rubenstein committed:
Adds google project display name and tests

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2125,6 +2130,7 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
               case Some(ba) if !billingProject.invalidBillingAccount => ba
               case _ => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Billing Account is missing or invalid for Billing Project: ${billingProject}"))
             }
+            val workspaceName = WorkspaceName(workspaceRequest.namespace, workspaceRequest.name)
```
maybe outside of the scope of this ticket, but the case class `WorkspaceName` seems ambiguous, especially when one of the params is the workspace name. I wonder if we can rename the case class to FullWorkspaceName or at least rename the val :/

gpolumbo-broad replied:
@andy7i, yeah, certainly outside the scope of this ticket.  I'm sure this is used in a ton of places but I 100% agree with you that the name is confusing and it throws me off all the time and I'd LOVE to change it.  Can you make a tech debt ticket?

andy7i replied:
yeah definitely, I didn't realize it but `WorkspaceName` is used in 100 places  -- created https://broadworkbench.atlassian.net/browse/CA-1264

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2134,7 +2140,7 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
               policyEmailsByName: Map[SamResourcePolicyName, WorkbenchEmail] = resource.accessPolicies.map(x => SamResourcePolicyName(x.id.accessPolicyName) -> WorkbenchEmail(x.email)).toMap
               (googleProjectId, googleProjectNumber) <- traceDBIOWithParent("setupGoogleProject", parentSpan)(_ => DBIO.from(
-                  setupGoogleProject(billingProject, billingAccount, workspaceId, policyEmailsByName, billingProjectOwnerPolicyEmail, parentSpan)))
+                  setupGoogleProject(billingProject, billingAccount, workspaceId, workspaceName, policyEmailsByName, billingProjectOwnerPolicyEmail, parentSpan)))
```
are there changes updating `setupGoogleProject`?

andy7i commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -1364,6 +1364,26 @@ class WorkspaceServiceSpec extends AnyFlatSpec with ScalatestRouteTest with Matc
+
+    val workspace = Await.result(services.workspaceService.createWorkspace(workspaceRequest), Duration.Inf)
+
+    verify(services.gcsDAO).updateGoogleProjectName(any[GoogleProjectId], ArgumentMatchers.eq(s"${testData.testProject1Name.value}/${newWorkspaceName}"))
```
suggestion: create a `val namespace = testData.testProject1Name.value`

andy7i commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -1364,6 +1364,26 @@ class WorkspaceServiceSpec extends AnyFlatSpec with ScalatestRouteTest with Matc
+
+    val workspace = Await.result(services.workspaceService.createWorkspace(workspaceRequest), Duration.Inf)
+
+    verify(services.gcsDAO).addGoogleProjectLabel(any[GoogleProjectId], ArgumentMatchers.eq("workspaceNamespace"), ArgumentMatchers.eq(newWorkspaceName))
+    verify(services.gcsDAO).addGoogleProjectLabel(any[GoogleProjectId], ArgumentMatchers.eq("workspaceName"), ArgumentMatchers.eq(testData.testProject1Name.value))
```
the namespace and name values are swapped

```suggestion
    verify(services.gcsDAO).addGoogleProjectLabel(any[GoogleProjectId], ArgumentMatchers.eq("workspaceNamespace"), ArgumentMatchers.eq(testData.testProject1Name.value))
    verify(services.gcsDAO).addGoogleProjectLabel(any[GoogleProjectId], ArgumentMatchers.eq("workspaceName"), ArgumentMatchers.eq(newWorkspaceName))
```

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
+    }
+  }
+
+  override def updateGoogleProjectName(googleProject: GoogleProjectId, name: String): Future[Unit] = {
```
suggestion: rename or add comment that this is the google project display name

gpolumbo-broad replied:
I have a slight preference towards commenting.  Since this is the `GoogleServicesDAO`, it is specifically about getting/setting values in Google land and this _is_ what the field is called on the Google Project.  That said, any time I need to do something with the Google Project Name or ID I reference this page:  https://cloud.google.com/resource-manager/docs/creating-managing-projects#before_you_begin which disambiguates `Project ID`, `Project Name`, and `Project Number`.  Some commenting here would be helpful and perhaps a link to the Google documentation.  Yeah, it might move, but it's the canonical source for what these things are and what they mean right now, and that's the best we can do.

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
+    val cloudResourceManagerServiceAccountCredential = getCloudResourceManagerServiceAccountCredential
+    val cloudResourceManager = getCloudResourceManager(cloudResourceManagerServiceAccountCredential)
+    for {
+      _ <- executeGoogleRequestWithRetry(cloudResourceManager.projects().update(googleProject.value, new Project().setName(name)))
```
It looks like we need to do some string validation because the displayname looks pretty limited :/

```
  /**
   * The user-assigned display name of the Project. It must be 4 to 30 characters. Allowed
   * characters are: lowercase and uppercase letters, numbers, hyphen, single-quote, double-quote,
   * space, and exclamation point.
   *
   * Example: My Project Read-write.
   * @param name name or {@code null} for none
   */
  public Project setName(java.lang.String name) {
```

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
     }
   }
 
+  override def addGoogleProjectLabel(googleProject: GoogleProjectId, key: String, value: String): Future[Unit] = {
```
do we do any string validation for key, value? some prior code example we did in DM: https://github.com/broadinstitute/gcp-dm-templates/blob/master/firecloud_project.py#L371

and buffer: https://github.com/DataBiosphere/terra-resource-buffer/blob/17f08a85c249546732975a998f3c7d261fa13e02/src/main/java/bio/terra/buffer/service/resource/flight/CreateProjectStep.java#L116

andy7i replied:
added validation ; waiting on answers from PO about how this should look https://broadworkbench.atlassian.net/browse/CA-1224

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
+    val cloudResourceManager = getCloudResourceManager(cloudResourceManagerServiceAccountCredential)
+    for {
+      project <- getGoogleProject(googleProject)
+      _ = project.getLabels.put(key, value)
```
just double checking if this does save the label to the map within the `project` object. Seems like it's a java map and I don't know what the behavior would be OTTOMH.

andy7i replied:
Not sure if this was working but I changed the logic to use more scala.

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
+    implicit val service: GoogleInstrumentedService.Value = GoogleInstrumentedService.CloudResourceManager
+
+    val cloudResourceManagerServiceAccountCredential = getCloudResourceManagerServiceAccountCredential
+    val cloudResourceManager = getCloudResourceManager(cloudResourceManagerServiceAccountCredential)
```
perhaps a nit, but it looks like these two lines of code are used 3 times now in `addGoogleProjectLabel`, `updateGoogleProjectName`, and `deleteGoogleProject`.  It's only 2 lines of code, so it's not that big of a deal, but we could consider extracting.

andy7i replied:
extracted as suggested and replaced it in some other areas where `getCloudResourceManager` was being used (sometimes with a SA that wasn't the CloudResourceManager SA)

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
     }
   }
 
+  override def addGoogleProjectLabel(googleProject: GoogleProjectId, key: String, value: String): Future[Unit] = {
```
Could we refactor this to take a `Map[String, String]` of labels so that we can add several at once and only make 1 call to Google?

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1030,46 @@ class HttpGoogleServicesDAO(
+    val cloudResourceManager = getCloudResourceManager(cloudResourceManagerServiceAccountCredential)
+    for {
+      project <- getGoogleProject(googleProject)
+      _ = project.getLabels.put(key, value)
```
I know I can probably Google this myself too and perhaps I will, but why for updating the labels, do we need to first call `getGoogleProject`, update the labels, and then `update` the project.  Whereas in `updateGoogleProjectName`, we do not get the project first and instead just create a new `Project` and specify the field we want updated?

andy7i replied:
RBS gives us projects with existing labels, so I think this is the right call so we don't replace any existing labels.  I've added a comment about this to clarify

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1927,6 +1928,10 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+      _ <- traceWithParent("updateGoogleProjectDisplayName", span)(_ => gcsDAO.updateGoogleProjectName(googleProjectId, workspaceName.toString))
+      _ <- traceWithParent("addGoogleProjectLabel: namespace", span)(_ => gcsDAO.addGoogleProjectLabel(googleProjectId, "workspaceNamespace", workspaceName.namespace))
+      _ <- traceWithParent("addGoogleProjectLabel: name", span)(_ => gcsDAO.addGoogleProjectLabel(googleProjectId, "workspaceName", workspaceName.name))
+      _ <- traceWithParent("addGoogleProjectLabel: id", span)(_ => gcsDAO.addGoogleProjectLabel(googleProjectId, "workspaceId", workspaceId))
```
As noted above, I think we can consolidate all the label calls into a single call.  Wondering if we can't also consolidate all of the calls to update the Google Project into a single call.  So set the Billing Account, Name, Labels all into a single `updateGoogleProject` call.  

Taking it a step further, `addGoogleProjectLabel` calls `getGoogleProject` as does `getGoogleProjectNumber`.  Could we optimize to make one call to get the Google Project (and therefore the Google Project Number) and then we can use that instance to make any updates to set the Labels, Billing Account, and Name, and then make one single call back to Google to update the Google Project.  

I think reducing the amount of chatter back and forth with Google is something we should generally aim for whenever we can to reduce overall latency, potential quota issues, and just overhead in general.

andy7i replied:
Combined the calls that I could, reducing ~7 calls to 2. 

The billing account uses a different service (cloud billing) so that can't be combined with the other calls here. Just in case, I did some investigation, but didn't see a way to batch calls to google across different gcp services (for example, cloud billing APIs vs cloud resource manager APIs).

If we need to, I think there could be some additional optimization for gcsDAO.removePolicyBindings and updateGoogleProjectIam to combine some of the calls, but it get messy so we discussed not spending too much time on that.

s-rubenstein replied:
Yeah, I had been thinking about that. I think it is also safer from a management perspective (what happens if an asynchronous call futzes with the labels.) Was thinking it would be cleanest just to have an `updateGoogleProject` call, but for some reason that wasn't the current pattern in the codebase, so I tried to stick with what was existing. Much happier combining though.

andy7i committed:
first pass

andy7i committed:
getCloudResourceManagerWithCloudResourceManagerServiceAccount()

andy7i committed:
if we care about # of items in the map (which I don't think we really do), we can do this instead:

```
    import org.mockito.ArgumentCaptor
    val mapSize = 3
    val captor = ArgumentCaptor.forClass(classOf[Map[String, String]])
    verify(services.gcsDAO).addGoogleProjectLabels(any[GoogleProjectId], captor.capture())
    assert(mapSize == captor.getValue.asInstanceOf[Map[String, String]].size)
```

andy7i committed:
addGoogleProjectLabels all at once instead of in multiple calls

andy7i committed:
combine calls to google (cloud resource manager)

andy7i committed:
combine calls to google (cloud resource manager) (pt 2 - missed this in the previous commit)

andy7i committed:
update tests

andy7i committed:
add labelSafeMap

andy7i committed:
unit test setUpProjectInCloudResourceManagerAndGetGoogleProjectNumber

andy7i committed:
Revert "unit test setUpProjectInCloudResourceManagerAndGetGoogleProjectNumber"

This reverts commit 7057654b

andy7i committed:
use getCloudResourceManagerWithCloudResourceManagerServiceAccount -- some uses of `getCloudResourceManager` weren't using the CRM SA

andy7i committed:
remove now-unused functions

andy7i committed:
add comments for updateGoogleProject

andy7i committed:
make and use getGoogleProjectNumberFromOption

andy7i committed:
might as well use googleProjectNameSafeString

andy7i committed:
remove now-unused functions pt 2 -- cleanup

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1927,7 +1930,8 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       googleProjectId <- traceWithParent("getGoogleProjectFromBuffer", span)(_ => resourceBufferService.getGoogleProjectFromBuffer(projectPoolType, workspaceId))
       _ <- traceWithParent("updateGoogleProjectBillingAccount", span)(_ => gcsDAO.updateGoogleProjectBillingAccount(googleProjectId, Option(billingAccount)))
-      googleProjectNumber <- traceWithParent("getProjectNumberFromGoogle", span)(_ => getGoogleProjectNumber(googleProjectId))
+      googleProjectLabels = Map("workspaceNamespace" -> workspaceName.namespace, "workspaceName" -> workspaceName.name, "workspaceId" -> workspaceId)
```
Opt: I think my preference would be to call labelSafeString on these here, but I am okay with it as is atm. I don't like that the `setUpProjectInCloudResourceManagerAndGetGoogleProjectNumber` changes labels around on users of the method. I would prefer that function to have minimal side effects, and that seems like a fairly obscure/hidden one.

andy7i replied:
I guess I was thinking that calling `labelSafeString()` outside of the function: 1) would require devs to know that there is a need to use `labelSafeString()` and to remember to use it before updating labels and 2) could result in obscure google error messages if we accidentally don't run `labelSafeString` and send an invalid label.

it does look a bit cleaner being outside the function so I updated to that

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1951,12 +1955,27 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+
+      // create a Project with fields that we want to update. Then send that to gcsDAO to update the project
+      googleProjectWithValuesToUpdate = new Project()
+        .setName(gcsDAO.googleProjectNameSafeString("Terra-managed Google project"))
```
TOL: This seems like minimal information here, would we want some sort of identifier on which Terra billing project it was associated with? I just wonder if we can give some information to users at all. Although maybe it doesn't matter. Do we expect users to care about their google project?

If so, who does care? Where will this be shown? I know we were going to show google project id in the workspace (in case users need that information for gs commands via the terminal or something), but I am wondering where this will really be used, especially since our users won't have direct access.

Maybe we/support are the users of this?

gpolumbo-broad replied:
I know we're waiting on feedback from Janki here, but whatever name we choose here, I think our preference should be to make it unique to each Workspace.

andy7i replied:
Janki and I went over the questions we had in the jira ticket and we decided to go with:

**display name**
* use the format “< namespace >--< name >” truncated to 30 chars

**Labels**
* use labels, truncating at 63 chars and replace all illegal characters with “-”

which should be reflected in the code now

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1927,7 +1930,8 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       _ <- traceWithParent("updateGoogleProjectBillingAccount", span)(_ => gcsDAO.updateGoogleProjectBillingAccount(googleProjectId, Option(billingAccount)))
-      googleProjectNumber <- traceWithParent("getProjectNumberFromGoogle", span)(_ => getGoogleProjectNumber(googleProjectId))
+      googleProjectLabels = Map("workspaceNamespace" -> workspaceName.namespace, "workspaceName" -> workspaceName.name, "workspaceId" -> workspaceId)
+      googleProjectNumber <- traceWithParent("setUpProjectInCloudResourceManagerAndGetGoogleProjectNumber", span)(_ => setUpProjectInCloudResourceManagerAndGetGoogleProjectNumber(googleProjectId, googleProjectLabels))
```
This feels a little clunky to me now.  Perhaps we try something like:
```suggestion
      googleProject <- traceWithParent("setUpProjectInCloudResourceManager", span)(_ => setUpProjectInCloudResourceManager(googleProjectId, googleProjectLabels))
      googleProjectNumber = gcsDAO.getGoogleProjectNumberFromOption(googleProjectId, Option(GoogleProjectNumber(googleProject.getProjectNumber.toString)))
```
Of course `setUpProjectInCloudResourceManager` would need to be updated to return the project, but that looks trivial.

My preference is to try not to have a single function doing too much.  The important thing this function is doing is the project setup.  The fact that it returns a Project or a Project Number is secondary, so l think this code becomes more maintainable if the steps are separated.

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -232,17 +233,55 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
+    name.toLowerCase.replaceAll("[^a-z0-9\\-'\" !]", "-").take(30)
+  }
+
+  def getGoogleProjectNumberFromOption(googleProjectId: GoogleProjectId, maybeGoogleProjectNumber: Option[GoogleProjectNumber]): GoogleProjectNumber = maybeGoogleProjectNumber match {
```
There's a couple of things I'm unsure about with this and I don't know what the answers are yet either unfortunately:

1. I'm not a huge fan of passing in the `googleProjectId` just for generating the error message.
2. I'm trying to ask myself what the true purpose is of this method.  Obviously it's DRY'ing up some code, that's good.  But the real "work" this method is doing for us doesn't seem to be "getting the Project Number", to me it seems to be about consistently handling the null case where our Project is missing the Number.  

I'm wondering if we could do something like `Option(googleProject.getProjectNumber).getOrElse(throw new RawlsExceptionWithErrorReport(...))`.  I think the things that I'd want to address are a consistent way of getting the `Long` project number off of the `googleProject`, turning it into a `GoogleProjectNumber`, and when it's null, generating the right ErrorReport.

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -232,17 +233,55 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
+
+  def getGoogleProjectNumberFromOption(googleProjectId: GoogleProjectId, maybeGoogleProjectNumber: Option[GoogleProjectNumber]): GoogleProjectNumber = maybeGoogleProjectNumber match {
+    case None => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadGateway, s"Failed to retrieve Google Project Number for Google Project ${googleProjectId}"))
+    case Some(longProjectNumber) => GoogleProjectNumber(longProjectNumber.toString)
```
```suggestion
    case Some(longProjectNumber) => longProjectNumber
```

andy7i committed:
yay unit tests! fix issues and tests

andy7i committed:
cleanup param

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -232,17 +233,60 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
+    * @param googleProject
+    * @return GoogleProjectNumber
+    */
+  def getGoogleProjectNumberFromGoogleProject(googleProject: Project): GoogleProjectNumber = googleProject.getProjectNumber match {
```
nit: I think it's ok to call this just `getGoogleProjectNumber` and it will read a just as well

andy7i committed:
`cloudResourceManager.projects().update` just returns the request back to us so return nothing instead

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/user/UserServiceSpec.scala:
```diff
@@ -88,11 +88,13 @@ class UserServiceSpec extends AnyFlatSpecLike with TestDriverComponent with Mock
       val folderId = "folders/1234567"
       when(mockGcsDAO.getFolderId(defaultServicePerimeterName.value.split("/").last)).thenReturn(Future.successful(Option(folderId)))
       when(mockGcsDAO.addProjectToFolder(project.googleProjectId, folderId)).thenReturn(Future.successful(()))
+      when(mockGcsDAO.getGoogleProjectNumberFromGoogleProject(googleProject)).thenReturn(googleProjectNumber)
```
I'm not sure you need to mock this?  I think that function should be self contained and should just work now?

andy7i replied:
I think because mockGcsDAO is a mockito mock:
```
val mockGcsDAO = mock[GoogleServicesDAO](RETURNS_SMART_NULLS)
```

it requires any call to be stubbed (even with my updates to `getGoogleProjectNumber`). Error message:

```
org.scalatest.exceptions.TestFailedException: The future returned an exception of type: org.mockito.exceptions.verification.SmartNullPointerException, with message: 
You have a NullPointerException here:
-> at org.broadinstitute.dsde.rawls.dataaccess.slick.WorkspaceComponent$workspaceQuery$.updateGoogleProjectNumber(WorkspaceComponent.scala:270)
because this method call was *not* stubbed correctly:
-> at org.broadinstitute.dsde.rawls.user.UserService.$anonfun$addProjectToServicePerimeter$9(UserService.scala:658)
googleServicesDAO.getGoogleProjectNumber(
    {"projectNumber" = 42L}
);
```

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -1031,20 +1029,35 @@ class HttpGoogleServicesDAO(
+    val cloudResourceManager: CloudResourceManager = getCloudResourceManagerWithCloudResourceManagerServiceAccount
+
+    for {
+      _ <- executeGoogleRequestWithRetry(cloudResourceManager.projects().update(googleProjectId.value, googleProjectWithUpdates))
```
I kind of agree with codecov here. It would give me a bit of confidence if we had a unit test here to know that we did in fact call `cloudResourceManager.projects().update()` but I don't see any other unit tests for GoogleServicesDAO. Maybe the right thing to do is to create an integration test?

s-rubenstein replied:
Yeah, I also looked for that… an integration test seems like a good idea.

andy7i committed:
PR feedback - setUpProjectInCloudResourceManager

andy7i committed:
pass the full google project with updates to `updateGoogleProject` and return the fully updated google project from `setUpProjectInCloudResourceManager` so that we aren't passing back the old, un-updated project.

not sure if there are any performance implications here but I don't see anything too extensive being passed as part of the update call.

andy7i reviewed (commented):
I can't comment on the earlier threads anymore.. 
I confirmed with janki about how the display name and labels should look
I also changed `getGoogleProjectNumberFromOption` but am considering inlining those instead like you were suggesting.

andy7i committed:
NPE check (thanks unit tests)

s-rubenstein commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -232,17 +233,60 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
+    * @return
+    */
+  def labelSafeMap(m: Map[String, String], prefix: String = ""): Map[String, String] = m.map { case (key, value) =>
+    labelSafeString(key, prefix) -> labelSafeString(value, prefix) }
```
Should this have the same default as `labelSafeString`? I am wondering if it will override the default and do it all with `""` as the prefix, which seems unintuitive to me.

s-rubenstein reviewed (commented):
Overall, looks good to me, with one comment. I can't hit approve cause I am the PR author.

andy7i committed:
pr feedback -- add default "fc-" prefix for `labelSafeMap` for consistency

andy7i reviewed (approved):
sky's 👍

gpolumbo-broad commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -1364,6 +1365,42 @@ class WorkspaceServiceSpec extends AnyFlatSpec with ScalatestRouteTest with Matc
+    verify(services.gcsDAO).updateGoogleProject(ArgumentMatchers.eq(GoogleProjectId("project-from-buffer")), captor.capture())
+    val capturedProject = captor.getValue.asInstanceOf[Project] // Explicit cast needed since Scala type interference and capturing parameters with Mockito don't play nicely together here
+
+    val expectedNewLabels = services.gcsDAO.labelSafeMap(Map("workspaceNamespace" -> newWorkspaceNamespace, "workspaceName" -> newWorkspaceName, "workspaceId" -> workspace.workspaceId), "")
```
I would just hard code the expected values here rather than using `labelSafeMap`.  I would also give the workspace name a couple of special characters so that you can make sure the `createWorkspace` call creates the labels with the appropriate special character replacement.  Maybe even make a long workspace name too to make sure it truncates properly.

andy7i committed:
integration test for updated labels and display name

andy7i committed:
update workbenchGoogleV from 0.21 to 0.21 -- should be non-breaking but is. Needs both the new dependency and exclusion rule to work.

andy7i committed:
workbenchGoogleV includes logstash dependency so the previous issue might be caused by too many logbacks in the kitchen

andy7i committed:
Merge branch 'v2_workspaces' into sr-add-google-project-display-name

andy7i committed:
pr feedback -- unit test to cover special characters, truncating, hardcoding the expected values.

gpolumbo-broad commented on automation/src/test/scala/org/broadinstitute/dsde/test/api/WorkspaceApiSpec.scala:
```diff
@@ -100,6 +100,40 @@ class WorkspaceApiSpec extends TestKit(ActorSystem("MySpec")) with AnyFreeSpecLi
         Rawls.billingV2.deleteBillingProject(billingProjectName)
       }
 
+      "to create a v2 workspaces (in a v2 billing project) with properly updated labels and display name" in {
```
Do we need this test?  This test suite already tests that we can create projects and `WorkspaceServiceSpec` already tests that the labels and display name are set correctly.  I don't see this test adding value or covering functionality that we don't already have covered.

If you _do_ think we need this test, then:

nit: for the sake of making progress, I don't think we should do this now, but I think this test should be named differently and located elsewhere.  I would rather call this test something like, "Rawls should set labels on the underlying Google Project when creating a new Workspace".  I feel like these clearly articulates the logic under test.  As it is currently written, it seems to me that this is a test about Rawls Billing Project owners and what they can do.  The fact that they can create Workspaces, to me, is one test.  The fact that when we create Workspaces, labels (and display name) are set is a totally different test.  

Again, it's a nit, no need to change it now, just something to think about as we create and organize tests.

andy7i replied:
I was able to set up the test name to what you suggested and the test seems happy. It seemed easy enough to move now so I opted to do that instead of waiting, in case we forgot.

(We talked about this in office time: I think we do need this test. The main difference between this test and the other tests is that this is an integration test. I think it provides value in that we don't really validate labels on a real project anywhere else. This test will ensure that we update real projects correctly. It saved me the hassle of finding the project in GCP console to verify if the project was updated correctly and should ensure that our logic for updating the project doesn't break in the future.)

andy7i committed:
pr feedback -- move and rename integration test

andy7i committed:
merged workbench libs change
